api/src/main/java/com/graphhopper/config/Profile.java

```diff
@@ -112,4 +113,8 @@ public boolean equals(Object o) {
     public int hashCode() {
         return name.hashCode();
     }
+
+    public int getVersion() {
+        return Helper.staticHashCode(toString());
```

Comment:
Hmm toString includes the hints (which is good/required for custom weighting), but unfortunately `PMap` uses an (unordered) `HashMap`, so this might go wrong? Shall we use `LinkedHashMap` in `PMap`? Or implement some method that returns the string in a deterministic way?

Reply:
Oh but I just realized that also `hints` does not even include the custom model but only the reference to the custom model file... Shouldn't our check also detect changes in the custom model file?

Reply:
> but unfortunately PMap uses an (unordered) HashMap, so this might go wrong

Ah, good catch!

> Oh but I just realized that also hints does not even include the custom model but only the reference to the custom model file... Shouldn't our check also detect changes in the custom model file?

Will write a test for this. I think we load the custom profile as object into this, but it might happen after the check.

Reply:
The CustomModel is stored in the hints already and produces a different hash code for mostly all situation, but relies on toString. So my test case passes. But only "mostly" because for areas we currently include only the count and so e.g. for changes of an area there would be no error. I will include the full areas and if this is too much at some point we can revert this to the count.

But the question is if we should rely on toString which should have only a debugging purpose? I think, yes, for now as the hashcode is not stable. And later work towards a config that is stored in the graph storage (and is loaded without providing it).

> Shall we use LinkedHashMap in PMap?

Yes, I'll do that (the other option would be TreeMap or putting toMap into a TreeMap but I think it cannot hurt to give PMap this ordering guarantuee)

Reply:
Cool thanks. We could have some method createContentString and call it from createHashCode **and** toString. This way its obvious that it has some meaning beyond debugging, we do not have to write it 'twice' and maybe we could even do the sorting there (in case we do not want to adjust PMap).

Reply:
Hmmh, sounds good but in PMap this still looks suboptimal:

```java
public String createContentString() {
    StringBuilder sb = new StringBuilder();
    for(Map.Entry<String, Object> entry : map.entrySet()) {
        sb.append(entry.getKey());
        sb.append('=');
        if(entry.getValue() instanceof ContentStringCreator)
            sb.append(((ContentStringCreator)entry.getValue()).createContentString()); // let e.g. CustomModel implement this interface
        else
            sb.append(entry.getValue().toString());
        sb.append(',');
    }
    return sb.toString();
}
```
WDYT?

Reply:
Ah, or do you mean to implement createContentString only in CustomModel as private method and use it in toString?

Reply:
> Ah, or do you mean to implement createContentString only in CustomModel as private method and use it in CustomModel.toString?

Yes